ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,177 @@
+	err := tmr.processTaskManifestMessage(message)
+	tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration).Done(err)
+	return err
+}
+
+// processTaskManifestMessage processes a task manifest message from ACS. It verifies
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum < agentLatestSeqNum {
+		tmr.logger.WithFields(logrus.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		}).Warn("Discarding task manifest message. " +
+			"Sequence number from manifest is less than Agent latest sequence number.")
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Get list of tasks on instance.
+	tasksToMessage, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(tmr, message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message asynchronously.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.ackTaskManifestMessage(message)
+	}()
+
+	// Send the task stop verification message to ACS asynchronously.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.sendTaskManifestMessageAck(message, tasksToMessage)
+	}()
+	wg.Wait()
+	return nil
+}
+
+func (tmr *taskManifestResponder) ackTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	messageID := aws.StringValue(message.MessageId)
+	tmr.logger.Debugf("Acking task manifest message id: %s", messageID)
+	err := tmr.respond(&ecsacs.AckRequest{
+		Cluster:           aws.String(tmr.cluster),
+		ContainerInstance: message.ContainerInstanceArn,
+		MessageId:         aws.String(messageID),
+	})
+	if err != nil {
+		tmr.logger.WithError(err).WithField("Error Acking TaskManifestMessage with messageID: %s", messageID)
+	}
+}
+
+func (tmr *taskManifestResponder) sendTaskManifestMessageAck(message *ecsacs.TaskManifestMessage, tasksToStop []*ecsacs.TaskIdentifier) {
```

Comment:
This method doesn't send an Ack but a TaskStopVerification, so how about renaming it to `sendTaskStopVerification` or something like that?